pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+
+        // Find the first schedule not due. All the schedule before `end` are due.
+        let now = Self::now_unix();
+        let end = schedules
+            .iter()
+            .position(|s| s.at > now) // Complexity `O(s)`.
+            .unwrap_or(schedules.len());
+
+        if end == 0 {
+            // Nothing found means no storage changes, so bail.
+            return Ok(());
+        };
+
+        // Ensure that ID count won't overflow.
+        // After this we're safe and can commit to storage.
+        let mut id_count = Self::next_ids(ticker, end as u64)?;
+
+        // Create checkpoints for due schedules.
+        let mut reschedule = Vec::new();
+        for schedule in schedules.drain(..end) {
+            if let Some(at) = schedule.schedule.next_checkpoint(now) {
+                // This schedule is recurring, so we'll need to reschedule.
+                reschedule.push(StoredSchedule { at, ..schedule });
+            }
+            let cp_id = id_count;
+            Self::create_at(None, *ticker, cp_id, schedule.at);
+            SchedulePoints::append((*ticker, schedule.id), cp_id);
+            id_count.0 += 1;
+        }
+
+        // Reschedule schedules we need to.
+        // Complexity: `O(r log s)`.
+        reschedule
+            .into_iter()
+            .for_each(|s| add_schedule(&mut schedules, s));
+
+        // Commit changes to `schedules`.
+        Schedules::insert(ticker, schedules);
+        Ok(())
+    }
+
+    /// Creates a schedule generating checkpoints
+    /// in the future at either a fixed time or at intervals.
+    pub fn create_schedule_base(
+        did: IdentityId,
+        ticker: Ticker,
+        schedule: CheckpointSchedule,
+        removable: bool,
+    ) -> Result<StoredSchedule, DispatchError> {
+        // Check the lower limit of the checkpoint period duration by computing the next
+        // checkpoint from the start of the schedule.
+        let oneshot = schedule.period.multiplier == 0;
+        ensure!(
+            oneshot
+                || schedule.next_checkpoint(schedule.start)
+                    >= Some(schedule.start + T::MinCheckpointDurationSecs::get()),
+            Error::<T>::ScheduleDurationTooShort
+        );
+
+        // Ensure the total complexity for all schedules is not too great.
+        let schedules = Schedules::get(ticker);
+        schedules
+            .iter()
+            .map(|s| s.schedule.period.complexity())
+            .try_fold(schedule.period.complexity(), |a, c| a.checked_add(c))
+            .filter(|&c| c <= SchedulesMaxComplexity::get())
+            .ok_or(Error::<T>::SchedulesTooComplex)?;
+
+        // Charge the fee for checkpoint creation.
+        T::ProtocolFee::charge_fee(ProtocolOp::AssetCreateCheckpointSchedule)?;
+
+        // Compute next schedule ID.
+        let id = Self::next_schedule_id(&ticker)?;
+
+        // If start is now, create the first checkpoint immediately.
+        let now = Self::now_unix();
```

Comment:
`schedule.start == now` is basically the condition when someone tries to use the schedule functionality as `create_checkpoint`. This is included to avoid the wrong logic. As for the user predicting `now_unix()`, they should use `create_checkpoint` directly instead and they won't need to predict anything.

Reply:
Update from meeting: Decided to hack in 0 as a sentinel value in the base function to deal with minting time vs. submission time ("predicting what now is").